File: checkpassword-server.go
package main

import "crypto/hmac"
import "crypto/md5"
import "encoding/json"
import "fmt"
import "io"
import "io/ioutil"
import "log"
import "net/http"
import "os"
import "os/exec"
import "strconv"
import "syscall"
import "time"
// Debugging
//import "github.com/davecgh/go-spew/spew"
// HTTP Routing
import "github.com/gorilla/mux"
// DB
import "database/sql"
import _ "github.com/go-sql-driver/mysql"
// Crypt (unix_md5_crypt)
// import "github.com/GehirnInc/crypt"
// OTP
import "github.com/hgfischer/go-otp"
// Getopt
import "github.com/DavidGamba/go-getoptions"

const configdir = "/etc/qbox"

//type authcachedata struct {
//	uid	int64
//	gid	int64
//	user	string
//	home	string
//	qboxuid	int64
//	qboxgid	int64
//}
//
//type authcachemeta struct {
//	epoch	int64
//	data	authcachedata
//}

type authfaildata struct {
	epoch	int64
	message	string
}

var authfail = make(map[string][]authfaildata)
var maxfail int
var failscript string

type clientreqdata struct {
	Username	string
	Password	string
	Timestamp	string
	Service		string
	Source		string
}

var db *sql.DB

func authenticate(w http.ResponseWriter, r *http.Request) {
	var services = [...]string {"smtp","smtps","pop","pops","pop3","pop3s","imap","imaps"}

	var authok bool = false
	var reqdata clientreqdata

	reqBody, err := ioutil.ReadAll(r.Body)
	if err != nil {
		w.WriteHeader(http.StatusBadRequest)
		fmt.Fprintf(w, "{\"error\":\"Could not read request body\"}\n")
		return
	}

	if len(reqBody) == 0 {
		// GET
		reqdata.Username  = mux.Vars(r)["username"]
		reqdata.Password  = mux.Vars(r)["password"]
		reqdata.Service   = mux.Vars(r)["service"]
		reqdata.Source    = mux.Vars(r)["source"]
		reqdata.Timestamp = mux.Vars(r)["timestamp"]
//		if err != nil {
//			w.WriteHeader(http.StatusBadRequest)
//			fmt.Fprintf(w, "{\"error\":\"Could not parse parameters\"}\n")
//			return
//		}
	} else {
		// POST
		err := json.Unmarshal(reqBody, &reqdata)
		if err != nil {
			w.WriteHeader(http.StatusBadRequest)
			fmt.Fprintf(w, "{\"error\":\"Could not parse JSON\"}\n")
			return
		}
	}

	// Check if serice is supported
	var servicesupported bool = false
	for _, v := range(services) {
		if v == reqdata.Service { servicesupported = true }
	}
	if !servicesupported {
		w.WriteHeader(http.StatusBadRequest)
		fmt.Fprintf(w, "{\"error\":\"Service '%s' is not supported\"}\n", reqdata.Service)
		return
	}

	// `root` is always denied
	if (reqdata.Username == "root") {
		fmt.Fprintf(os.Stderr, "User root denied on %s from %s\n", reqdata.Service, reqdata.Source)
		w.WriteHeader(http.StatusForbidden)
		fmt.Fprintf(w, "{\"error\":\"root logins are prohibited\"}\n")
		return
	}

	// Check that DB is still there
	err = db.Ping()
	if err != nil {
		w.WriteHeader(http.StatusServiceUnavailable)
		fmt.Fprintf(w, "{\"error\":\"Database unavailable\"}\n")
		return
	}

	// Query DB
	type dbschema struct {
		password	string
		homedir		string
		sysuid		int64
		sysgid		int64
		quota		int64
		uid		int64
		gid		int64
		oathtoken	string
		aliasof		string
	}
	var dbdata dbschema

	// Prepare and execute query
	stmt1, err := db.Prepare("SELECT password,homedir,sysuid,sysgid,quota,uid,gid,oath_token,alias_of FROM passwd WHERE username = ? AND ? != '' limit 1")
	if err != nil {
		fmt.Println("Prepare SELECT FROM passwd failed: "+err.Error())
	}
	rows1, err := stmt1.Query(reqdata.Username, reqdata.Service)
	if err != nil {
		fmt.Println("Executing SELECT FROM passwd failed: "+err.Error())
	}
	defer stmt1.Close()

	// Read query results
	for rows1.Next() {
		err := rows1.Scan(&dbdata.password,
				  &dbdata.homedir,
				  &dbdata.sysuid,
				  &dbdata.sysgid,
				  &dbdata.quota,
				  &dbdata.uid,
				  &dbdata.gid,
				  &dbdata.oathtoken,
			          &dbdata.aliasof)
		if err != nil {
			fmt.Println("Scaning SELECT FROM passwd result failed: "+err.Error())
		}
	}
	rows1.Close()

	if (reqdata.Password == dbdata.password) { authok = true }
	// CRAM-MD5
	if (reqdata.Password == hmac_md5_hex(reqdata.Timestamp, dbdata.password)) { authok = true }
	// APOP
	if (reqdata.Password == md5_hex(reqdata.Timestamp, dbdata.password)) { authok = true }
	// OTP
	if (otp_verify(dbdata.oathtoken, reqdata.Password)) { authok = true }

	if authok {
		// Write to log
		fmt.Fprintf(os.Stderr, "Authentication succeeded for %s on %s from %s\n", reqdata.Username, reqdata.Service, reqdata.Source);

		// Support aliasing
		if (dbdata.aliasof != "") {
			stmt2, err := db.Prepare("SELECT password,homedir,sysuid,sysgid,quota,uid,gid,oath_token,alias_of FROM passwd WHERE username = ? limit 1")
			if err != nil {
				fmt.Println("Prepare SELECT FROM passwd failed for alias: "+err.Error())
			}
			rows2, err := stmt2.Query(dbdata.aliasof)
			if err != nil {
				fmt.Println("Executing SELECT FROM passwd failed for alias: "+err.Error())
			}
			defer stmt2.Close()

			for rows2.Next() {
				err := rows2.Scan(&dbdata.password,
						  &dbdata.homedir,
						  &dbdata.sysuid,
						  &dbdata.sysgid,
						  &dbdata.quota,
						  &dbdata.uid,
						  &dbdata.gid,
						  &dbdata.oathtoken,
					          &dbdata.aliasof)

				if err != nil {
					fmt.Println("Scaning SELECT FROM passwd result failed for alias: "+err.Error())
				}
			}
			rows2.Close()
		}

		// Send response to checkpassword-client
		w.WriteHeader(http.StatusOK)
		rawin := json.RawMessage(`{"user":"`+reqdata.Username+`",`+
					 `"home":"`+dbdata.homedir+`",`+
					 `"uid":`+strconv.FormatInt(dbdata.sysuid, 10)+`,`+
					 `"gid":`+strconv.FormatInt(dbdata.sysgid, 10)+`,`+
					 `"qboxuid":`+strconv.FormatInt(dbdata.uid, 10)+`,`+
					 `"qboxgid":`+strconv.FormatInt(dbdata.gid, 10)+`}`)
		bytes, err := rawin.MarshalJSON()
		if err != nil {
			fmt.Println("Failed to marshal response: "+err.Error())
		}
		fmt.Fprintf(w, string(bytes))

		// Update `lastlogin` table
		if !update_lastlogin(dbdata.uid, reqdata.Username, reqdata.Service) {
			fmt.Println("Failed to update lastlogin table for "+reqdata.Username)
		}
	} else {
		// Write to log
		var logline string
		if dbdata.uid > 0 {
			logline = fmt.Sprintf("Authentication failed for %s on %s from %s [password: %s]\n", reqdata.Username, reqdata.Service, reqdata.Source, reqdata.Password)
		} else {
			logline = fmt.Sprintf("User %s unknown on %s from %s [password: %s]\n", reqdata.Username, reqdata.Service, reqdata.Source, reqdata.Password)
		}
		fmt.Fprint(os.Stderr, logline)

		// Send response to checkpassword-client
		w.WriteHeader(http.StatusForbidden)

		// Record auth failure for later
		authfail[reqdata.Source] = append(authfail[reqdata.Source], authfaildata{epoch: time.Now().Unix(), message: timestamp()+` - `+logline} )

		// If maximum authentication failures are reached, call failscript
		if len(authfail[reqdata.Source]) >= maxfail &&
		   len(failscript) > 0 {
                        // We copy the data to a temporary structure to prevent
                        // a race-condition. If a client is trying rapidly, the
                        // failscript could otherwise be called multiple times
                        var authhistory []authfaildata = authfail[reqdata.Source]
			delete(authfail, reqdata.Source)

			fmt.Println("Calling "+failscript+" for "+reqdata.Source)
			cmd := exec.Command(failscript, reqdata.Source)
			cmdstdin, err := cmd.StdinPipe()
			if err != nil {
				fmt.Println("Failed to connect to stdin: ", err)
			}
			err = cmd.Start()
			if err != nil {
				fmt.Println("Failed to run failscript: ", err)
			}
			//for _, v := range authfail[reqdata.Source] {
			for _, v := range authhistory {
				_, err := cmdstdin.Write([]byte(v.message))
				if err != nil {
					fmt.Println("Failed to write to stdin: ", err)
				}
			}
			cmdstdin.Close()
			err = cmd.Wait()
			if err != nil {
				fmt.Println("Failed to wait for failscript: ", err)
			}
			//delete(authfail, reqdata.Source)
			// ^^ moved to top
		}
	}
}

func fileExists(filename string) bool {
        info, err := os.Stat(filename)
        if os.IsNotExist(err) {
                return false
        }

        return !info.IsDir()
}

func main () {
	// Option parsing
	var listenport int
	opt := getoptions.New()
	opt.IntVar(&listenport, "port", 17520)
	opt.IntVar(&maxfail, "maxfail", 10)
	opt.StringVar(&failscript, "failscript", "")
	_, parseerr := opt.Parse(os.Args[1:])
	if parseerr != nil {
		fmt.Print(opt.Help())
		log.Fatal(parseerr)
	}

	// Check that failscript exists
	if failscript != "" {
		if _, err := os.Stat(failscript); os.IsNotExist(err) {
			log.Fatal(err)
		}
	}

	// Running as root is discouraged
	if (syscall.Getuid() == 0) {
		fmt.Println("Running as root is not supported!")
		os.Exit(2)
	}

	// Read config files
        var dbserver string = "127.0.0.1"
        if fileExists(configdir + "/dbserver") {
                buf, err := ioutil.ReadFile(configdir + "/dbserver")
                if err == nil {
                        dbserver = string(buf)
                }
        }

        var dbuser string = "qbox"
        if fileExists(configdir + "/dbuser") {
                buf, err := ioutil.ReadFile(configdir + "/dbuser")
                if err == nil {
                        dbuser = string(buf)
                }
        }

        var dbpass string
        if fileExists(configdir + "/dbpass") {
                buf, err := ioutil.ReadFile(configdir + "/dbpass")
                if err == nil {
                        dbpass = string(buf)
                }
        }

	var err error
        db, err = sql.Open("mysql", dbuser+":"+dbpass+"@tcp("+dbserver+")/qbox")
	if err == nil {
		err = db.Ping()
		if err != nil {
			log.Fatal(err)
			os.Exit(1)
		}
		defer db.Close()
	}

	router := mux.NewRouter().StrictSlash(true)
	router.HandleFunc("/", authenticate).Methods("POST")
        router.HandleFunc("/{service}/{username}/{password}/{timestamp}/{source}", authenticate).Methods("GET")
	log.Fatal(http.ListenAndServe("127.0.0.1:"+strconv.FormatInt(int64(listenport), 10), router))
}

func hmac_md5_hex(salt string, password string) string {
	hash := hmac.New(md5.New, []byte(password))
	_, err := io.WriteString(hash, salt)
	if err != nil {
		fmt.Println("Failed to write to hash function: ", err)
	}

	return fmt.Sprintf("%x", hash.Sum(nil))
}

func md5_hex(timestamp string, password string) string {
	hash :=	md5.New()
	_, err := io.WriteString(hash, timestamp + password)
	if err != nil {
		fmt.Println("Failed to write to hash function: ", err)
	}

	return fmt.Sprintf("%x", hash.Sum(nil))
}

// Unix authentication is not supported so the below is dead code
//func unix_md5_crypt(salt string, password string) string {
//        crypt := crypt.MD5.New()
//        ret, _ := crypt.Generate([]byte(password), []byte(salt))
//
//        return ret
//}

func otp_verify(token string, password string) bool {
	totp := otp.TOTP{Secret: token, IsBase32Secret: true, WindowBack: 20, WindowForward: 20}
	return totp.Verify(password)
}

func update_lastlogin(uid int64, username string, service string) bool {
	stmt, err := db.Prepare("INSERT INTO lastlogin VALUES (?, ?, UNIX_TIMESTAMP(NOW()), TIMESTAMP(NOW()), ?) ON DUPLICATE KEY UPDATE epoch=UNIX_TIMESTAMP(NOW()), timestamp=TIMESTAMP(NOW()), protocol=?")
	if err != nil {
		fmt.Println(err)
		return false
	}
	defer stmt.Close()

	_, err = stmt.Exec(strconv.FormatInt(uid, 10), username, service, service)
	if err != nil {
		fmt.Println(err)
		return false
	} else {
		return true
	}
}

func timestamp() string {
	const timelayout = "2006-01-02 15:04:05"
	return time.Unix(time.Now().Unix(), 0).Format(timelayout)
}

File: legacy/checkpassword-server.pl
#!/usr/bin/perl -T

use strict;
use warnings;

use Carp;
use DBI;
use DBD::mysql;
use User::pwent;
use Crypt::PasswdMD5;
use Dancer ':script';
use English qw(-no_match_vars);
use Getopt::Long;
use MIME::Base64;
use Digest::MD5 qw(md5 md5_hex md5_base64);
use Digest::HMAC_MD5 qw(hmac_md5 hmac_md5_hex);
# Logging
use File::Basename qw(basename);
#use Sys::Syslog;
use POSIX qw/strftime setuid setgid/;
# Only for OATH
use Authen::OATH;
use Time::Local;
use JSON ();

my ($port, $maxfail, $runas, $failscript);
my $cachetime = 3600;
GetOptions("port=i"       => \$port,
           "maxfail=i"    => \$maxfail,
           "user=s"       => \$runas,
	   "failscript=s" => \$failscript,
           "cachetime=s"  => \$cachetime);

# Disable output buffering
*STDOUT->autoflush();
*STDERR->autoflush();

local $ENV{PATH} = '';

my $configdir = "/etc/qbox";

my ($dbpassword, $salt, $cryptpw, $md5pw) = ('','','','');
my $dummy;
my ($homedir, $uid, $gid, $shell, $quota, $qboxuid, $qboxgid, $oathtoken);
my $sqlrows;
my ($db,$sth,$dbh,$statement,$success);
my %authfail;
my $epoch;
my $runasuser;

#  Drop privileges
if (defined($runas)) {
  $runasuser = getpwnam($runas);
  if (not(($runasuser->uid =~ /\d+/x) && ($runasuser->gid =~ /\d+/x))) {
    die("Could not determine UID/GID for $runas\n");
  }
  POSIX::setgid($runasuser->gid);
  POSIX::setuid($runasuser->uid);
  info "Droped privileges to $runas (gid $GID egid $EGID uid $UID euid $EUID)\n";
} else {
  warn "Still running with full privileges\n";
}

# Auth cache
my %authcache;

&connect_mysql() || die "Could not connect to database: $!\n";

# We are working with JSON
set serializer => 'JSON';

# Logger
set logger => 'console';

# TCP Port
if (not(defined($port))) { $port = 7520 }
set port => $port =~ /(\d+)/ix;
  
# Collect login data
# ******************
get '/:service/:username/:password/:timestamp/:source' => sub { &authenticate };
post '/' => sub { &authenticate };
get '/admin/authcache' => sub { &dump_authcache };

dance;
exit;

sub authenticate {

  my $service = param('service');
  my $username = param('username');
  my $password = param('password');
  my $ipaddr = param('source');
  my $timestamp = param('timestamp');

  # Check auth cache
  if (defined($authcache{$username}{$password})) {
    if ($authcache{$username}{$password}{'epoch'} > (time - $cachetime)) {
      print STDERR "Authentication succeeded for $username on $service from $ipaddr [cached]\n";
      return { %{$authcache{$username}{$password}{'data'}} };
    } else {
      print STDERR "Authcache entry for $username is too old (".$authcache{$username}{$password}{'epoch'}." vs ".time.")\n";
    }
  }

  # Is this service/protocol supported
  ####################################
  my @services = ('pop', 'smtp', 'imap');
  if (not(grep { /$service/ } @services)) {
    status 400;
    return;
  }

  # Do not allow root logins
  # ************************
  if ($username eq 'root') {
    print STDERR "User root denied on $service from $ipaddr\n";
    status 403;
    return;
  }

  # Check that MySQL is still there
  #################################
  if (not($dbh->ping)) {
    # connect again
    print STDERR "Database has gone away. Reconnecting\n";
    &connect_mysql;
  }
  
  # Get the authenticating users data from MySQL
  # ********************************************
  &get_mysqldata();
  if ($sqlrows == 0) {
    # Read /etc/passwd if SQL result is empty
    &get_unixdata($username);
  }

  # Abort if no password or homedir is defined
  # ******************************************
  if ( (not(defined($dbpassword))) or (not(defined($homedir))) ) {
    print STDERR "User unknown for $username on $service from $ipaddr [password: $password]\n";
    $epoch = time;
    $authfail{$ipaddr}{$epoch} = &timestamp." - User unknown for $username on $service from $ipaddr [password: $password]\n";
    &lockout($ipaddr);
    status 403;
    return;
  }

  # Handle crypted passwords
  # ************************
  if ((length($dbpassword)) == 13) {
    if ($dbpassword =~ /^(.{2})/x) { $salt = $1 }
    $cryptpw = crypt($password, $salt);
  }

  # Handle MD5 passwords
  # ********************
  if ((length($dbpassword)) == 34) {
    if ($dbpassword =~ /^\$1\$(.{8})\$/x) { $salt = $1 }
    $md5pw = unix_md5_crypt($password, $salt);
  }

  # Handle CRAM-MD5 authentication (SMTP AUTH)
  # ******************************************
  if (($password =~ /^[0-9a-f]+$/x) && ($service eq 'smtp')) {
#   debug("SMTP AUTH detected\n");
    $dbpassword = hmac_md5_hex($timestamp, $dbpassword);
  }

  # Handle APOP authentication (POP3)
  # *********************************
  if ((length($password) == 32) && ($service eq 'pop')) {
#   debug("APOP detected\nEncoding: $timestamp$dbpassword\n");
    $dbpassword = md5_hex($timestamp,$dbpassword);
  }

  # Handle OATH authentication
  # **************************
  my $oathdrift = 180;
  my $oathok = 0;
     $epoch = timelocal(localtime);
  my $oath = Authen::OATH->new();

  if (defined($oathtoken)) {
    foreach my $x (($epoch - $oathdrift) .. ($epoch + $oathdrift)) {
      if (($x % 30) == 0) {
        if ( ($oath->totp($oathtoken, $x)) eq $password ) {
#         debug("OATH Token successful.");
          $oathok = 1;
        }
      }
    }
  }

  # Compare the supplied password to the stored data
  # ************************************************
  if ( ($password eq $dbpassword) || ($cryptpw eq $dbpassword) || ($md5pw eq $dbpassword) || ($oathok == 1) ) {
    print STDERR "Authentication succeeded for $username on $service from $ipaddr\n";
    &update_lastlogin($qboxuid, $username, $service);
    status 200;
    undef($authfail{$ipaddr});
    $authcache{$username}{$password}{'epoch'} = time;
    $authcache{$username}{$password}{'data'}  = { home => $homedir, user => $username, uid => $uid+0, gid => $gid+0, qboxuid => $qboxuid+0, qboxgid => $qboxgid+0 };
    return { home => $homedir, user => $username, uid => $uid+0, gid => $gid+0, qboxuid => $qboxuid+0, qboxgid => $qboxgid+0 };
  }

  # Abort on failed authentication
  # ******************************
  else {
    print STDERR "Authentication failed for $username on $service from $ipaddr [password: $password]\n";
    $epoch = time;
    $authfail{$ipaddr}{$epoch} = &timestamp." - Authentication failed for $username on $service from $ipaddr [password: $password]\n";
    &lockout($ipaddr);
    status 403;
    return;
  }
}


# Subs related to database operations
# ***********************************
sub connect_mysql {
  $db = "dbi:mysql:qbox; host=".&content("$configdir\/dbserver");
  $dbh = DBI->connect($db, &content("$configdir\/dbuser"), &content("$configdir\/dbpass"));

  if ($dbh) {
    return 1;
  }
  return 0;
}

sub disconnect_mysql {
  $sth->finish;
  $dbh->disconnect;
  return;
}

sub make_query {
  $sth = $dbh->prepare($statement);
  $success = $sth->execute();
  $sqlrows = $sth->rows();
  return;
}
# ***********************************

# Subs related to fetching user data
# **********************************
sub get_mysqldata {
  $statement = "select password,homedir,sysuid,sysgid,quota,uid,gid,oath_token from passwd where username = \"".param('username')."\" and ".param('service')." != '' limit 1";
  &make_query();
  ($dbpassword,$homedir,$uid,$gid,$quota,$qboxuid,$qboxgid,$oathtoken) = $sth->fetchrow_array();
  return;
}

sub get_unixdata {
  my $username = shift;
  while($dummy = getpwent()) {
    if (($dummy->name) eq $username) {
      $homedir = $dummy->dir;
      $uid = $dummy->uid;
      $gid = $dummy->gid;
      $shell = $dummy->shell;
      $dbpassword = $dummy->passwd;
      $quota = 0;
    }
  }
  return;
}
# **********************************

sub content {
  my ($file) = @_;
  if (-r $file) {
    open(my $FILEFH, "<", $file) || croak "Could not read $file";
    my $input = <$FILEFH>;
    chomp($input);
    close($FILEFH);
    return $input;
  } else {
    return '';
  }
}

sub lockout {
  my ($ip) = @_;

  # Untaint IP
  if ($ip =~ /(\d+\.\d+\.\d+\.\d+)/ix) {
    $ip = $1;
  } else {
    return 0;
  }
  
  if (defined($maxfail)) {
    if (keys($authfail{$ip}) >= $maxfail) {
      if ($failscript =~ m/(\/.*)/ix) { $failscript = $1; }
      print STDERR "Calling $failscript for $ip\n";
      open(my $FSCRIPT, "|-", "$failscript $ip") || croak "Could not invoke $failscript\n";
        foreach (sort(keys(%{$authfail{$ip}}))) {
           print $FSCRIPT $authfail{$ip}{$_};
        }
      close($FSCRIPT);
      undef($authfail{$ip});
    }
  }
  return 1;
}

sub timestamp {
  return strftime('%b %d %H:%M:%S', gmtime);
}

sub update_lastlogin {
  my ($lluid, $llusername, $lservice) = @_;
  $statement = 'INSERT INTO lastlogin VALUES ("'.$lluid.'","'.$llusername.'",UNIX_TIMESTAMP(NOW()),TIMESTAMP(NOW()),"'.$lservice.'") ON DUPLICATE KEY UPDATE epoch=UNIX_TIMESTAMP(NOW()), timestamp=TIMESTAMP(NOW()), protocol="'.$lservice.'"';
  &make_query();

  return;
}

sub dump_authcache {
  status 200;
  return { %authcache };
}
